eclair-core/src/main/scala/fr/acinq/eclair/message/OnionMessages.scala

```diff
@@ -58,24 +58,24 @@ object OnionMessages {
         .map(tlvs => RouteBlindingEncryptedDataCodecs.blindedRouteDataCodec.encode(TlvStream(tlvs)).require.bytes)
     }
     destination match {
-      case BlindedPath(route) if originKey.exists(route.introductionNodeId == _.publicKey) =>
-        val (nextNodeId, nextBlinding) = RouteBlindingEncryptedDataCodecs.decode(originKey.get, route.blindingKey, route.blindedNodes.head.encryptedPayload) match {
-          case Left(invalidData) => throw new Exception(invalidData.toString)
-          case Right(decoded) =>
-            (decoded.tlvs.get[RouteBlindingEncryptedDataTlv.OutgoingNodeId].get.nodeId,
-              decoded.tlvs.get[RouteBlindingEncryptedDataTlv.NextBlinding].map(_.blinding).getOrElse(decoded.nextBlinding))
+      case BlindedPath(route) if nodeKey.publicKey == route.introductionNodeId =>
+        RouteBlindingEncryptedDataCodecs.decode(nodeKey, route.blindingKey, route.blindedNodes.head.encryptedPayload) match {
+          case Left(f) => Failure(new IllegalArgumentException(f.message))
+          case Right(decoded) => decoded.tlvs.get[RouteBlindingEncryptedDataTlv.OutgoingNodeId].map(_.nodeId) match {
+            case Some(nextNodeId) => Success(Sphinx.RouteBlinding.BlindedRoute(nextNodeId, decoded.nextBlinding, route.blindedNodes.tail))
+            case None => Failure(new IllegalArgumentException("outgoing_node_id is missing from route blinding data"))
+          }
         }
-        Sphinx.RouteBlinding.BlindedRoute(nextNodeId, nextBlinding, route.blindedNodes.tail)
       case Recipient(nodeId, pathId, padding) =>
         val tlvs: Set[RouteBlindingEncryptedDataTlv] = Set(padding.map(Padding), pathId.map(PathId)).flatten
         val lastPayload = RouteBlindingEncryptedDataCodecs.blindedRouteDataCodec.encode(TlvStream(tlvs)).require.bytes
-        Sphinx.RouteBlinding.create(blindingSecret, intermediateNodes.map(_.nodeId) :+ nodeId, intermediatePayloads :+ lastPayload).route
```

Comment:
I don't agree that adding `Success` everywhere is cleaner.

Reply:
Removing code that throws exception is a no-brainer: we simply should never throw exceptions if we can avoid it.

Reply:
But why? The code is much nicer.

Reply:
On top of having bad performance because it captures a stack trace that we don't need, and needing to catch that exception just to actually convert it to a `Failure()` (which is what the `Try` does), it's a shortcut to avoid properly handling error cases: the error message we'll get for a `option.get` won't be helpful at all compared to an error message that says that the tlv stream doesn't contain a necessary outgoing node id. The code that throws exceptions is more concise, but it's definitely not nicer IMO...